### PR TITLE
[shim] minimal implementation for shim-runc-v2

### PR DESCRIPTION
**Pull request by Yuna-Tomi**
minimal implementation for containerd-shim-runc.
limitations are described in #21 

Signed-off-by: Yuna Tomida <ytomida.mmm@gmail.com>

### CODE REVIEW

**Comment by Yuna-Tomi** [comment]
Difference from #21:
- Stdio handling via pipe is suppported
- OOM event monitor is supported

File: crates/shim-runc/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+repository = "https://github.com/containerd/rust-extensions"
+description = "rust extensions for containerd-shim-runc"
+keywords = ["containerd", "shim", "containers", "ttrpc", "runc"]
+categories = ["api-bindings"]
+homepage = "https://containerd.io"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+cgroups-rs = { git =  "https://github.com/kata-containers/cgroups-rs" }
```
Comment (jiangliu) [inline crates/shim-runc/Cargo.toml:16]: Why using a git reference instead of published crate from crates.io?

File: crates/shim-runc/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+[package]
+name = "containerd-shim-runc"
+version = "0.1.0"
+authors = ["Yuna Tomida <ytomida.mmm@gmail.com>", "The containerd Authors"]
+edition = "2018"
+license = "Apache-2.0"
+repository = "https://github.com/containerd/rust-extensions"
+description = "rust extensions for containerd-shim-runc"
```
Comment (jiangliu) [inline crates/shim-runc/Cargo.toml:8]: "rust implementation of"?

Comment (Yuna-Tomi) [reply crates/shim-runc/Cargo.toml:8]: That's good, thank you!

File: crates/shim-runc/build.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+use std::fs;
+use std::path::Path;
+use ttrpc_codegen::{Codegen, ProtobufCustomize};
+
+fn main() {
```
Comment (jiangliu) [inline crates/shim-runc/build.rs:21]: How about applying the same trick as https://github.com/containerd/rust-extensions/blob/main/crates/shim-protos/Cargo.toml#L25 ?

File: crates/shim-runc/src/container.rs [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    init::InitProcess,
+};
+use crate::utils;
+
+const OPTIONS_FILENAME: &str = "options.json";
+
+#[derive(Debug)]
+/// Struct for managing runc containers.
+pub struct Container {
+    mu: Arc<Mutex<()>>,
```
Comment (jiangliu) [inline crates/shim-runc/src/container.rs:53]: Why mu is needed? The rust borrow checker should have already protected `Container` objects from concurrent access.

File: crates/shim-runc/src/container.rs [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+            rootfs: mounts.clone(),
+            terminal: req.terminal,
+            stdin: req.stdin.clone(),
+            stdout: req.stdout.clone(),
+            stderr: req.stderr.clone(),
+            options: req.options.clone().into_option(),
+        };
+
+        // Write options to file, which will be removed when shim stops.
+        match write_options(&req.bundle, &opts) {
```
Comment (jiangliu) [inline crates/shim-runc/src/container.rs:110]: Could be simplified as below?
```
write_options(&req.bundle, &opts)?;
```

**Comment by mxpv** [comment]
> Difference from https://github.com/containerd/rust-extensions/pull/21:
> - Stdio handling via pipe is suppported
> - OOM event monitor is supported

@Yuna-Tomi We got https://github.com/containerd/rust-extensions/pull/24 in. Would you be interested to fulfill the gaps from this PR?

**Comment by Yuna-Tomi** [comment]
@mxpv 
Sorry for late reply...
Thank you for recommendation!
The first feature (stdio handlings) seems to be already available, so I'll try latter one.

**Pull request closed without merging**